src/main/java/org/broadleafcommerce/payment/service/gateway/autoconfigure/PayPalServiceAutoConfiguration.java

```diff
@@ -0,0 +1,106 @@
+package org.broadleafcommerce.payment.service.gateway.autoconfigure;
+
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutConfiguration;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutConfigurationService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutExternalCallService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutHostedService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutReportingService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutRollbackService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutTransactionConfirmationService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalCheckoutTransactionService;
+import org.broadleafcommerce.payment.service.gateway.DefaultPayPalSyncTransactionService;
+import org.broadleafcommerce.payment.service.gateway.PayPalCheckoutConfiguration;
+import org.broadleafcommerce.payment.service.gateway.PayPalCheckoutExternalCallService;
+import org.broadleafcommerce.payment.service.gateway.PayPalSyncTransactionService;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.env.Environment;
+
+import com.broadleafcommerce.paymentgateway.service.configuration.PaymentGatewayConfigurationService;
+import com.broadleafcommerce.paymentgateway.service.hosted.PaymentGatewayHostedService;
+import com.broadleafcommerce.paymentgateway.service.reporting.PaymentGatewayReportingService;
+import com.broadleafcommerce.paymentgateway.service.rollback.PaymentGatewayRollbackService;
+import com.broadleafcommerce.paymentgateway.service.transaction.PaymentGatewayTransactionConfirmationService;
+import com.broadleafcommerce.paymentgateway.service.transaction.PaymentGatewayTransactionService;
+
+@Configuration
+public class PayPalServiceAutoConfiguration {
+
+    @Bean
+    @ConditionalOnMissingBean
+    public PayPalCheckoutExternalCallService externalCallPayPalCheckoutService(
+            PayPalCheckoutConfiguration configuration) {
+        return new DefaultPayPalCheckoutExternalCallService(configuration);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public PayPalCheckoutConfiguration payPalCheckoutConfiguration(Environment env) {
+        return new DefaultPayPalCheckoutConfiguration(env);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public PaymentGatewayConfigurationService paymentGatewayConfigurationService(
```

Comment:
I think that the original idea of the payment gateways was that you could have multiple payment gateways in a single project. Especially for Paypal; it is common to accept Paypal accounts and use a different provider for credit cards (maybe Stripe, idk).

However, the way you have this auto configuration you have effectively done this:

```java
@ConditionalOnMissingBean(PaymentGatewayConfigurationService.class)
```

as `@ConditionalOnMissingBean` uses the method return value to check if you do not provide a manual value  in the annotation itself. So, this means that if some other service produces a `PaymentGatewayConfigurationService` (which every single one will) then this bean will not be registered.

@elbertbautista can you clarify the requirements here on what needs to hold true for these payment gateways?

Reply:
It's also likely that we use some different patterns than we had before in the payment gateway service in regards to how these beans are used.

Reply:
@phillipuniverse great catch! It's definitely still expect that multiple gateways will be used within a single project. The pattern used to register these beans needs to be updated.